rustls/src/msgs/deframer.rs

```diff
@@ -96,6 +98,8 @@ impl MessageDeframer {
         // we do a zero length read.  That looks like an EOF to
         // the next layer up, which is fine.
         debug_assert!(self.used <= OpaqueMessage::MAX_WIRE_SIZE);
+        let need_capacity = Ord::min(OpaqueMessage::MAX_WIRE_SIZE, self.used + READ_SIZE);
```

Comment:
With this change we no longer read the largest amount possible that fits in the `OpaqueMessage::MAX_WIRE_SIZE` buffer; instead we read at most 4096 bytes at a time. This is because we'd like to not pre-emptively allocate the large buffer if it ends up not being needed (see #794) and 4096 bytes seems to be enough for most common cases. We could read multiple times on a full read, but that would make the return type more complicated, since we'd have to yield both the amount of bytes read and a potential error.

Reply:
One future tweak I had in mind is to reset the buffer size after the handshake is done, which should be responsible for most of the larger messages. But even now I'm seeing that a real-world handshake probably doesn't need to grow the buffer.

We also technically should allow the buffer to grow larger during the handshake (previously I think the joiner allowed up to 64k bytes).

Reply:
I've added some logic for this, see https://github.com/rustls/rustls/pull/1171/files#diff-7444beb5839fece40e64a80048a8cc03ab7b9f2e799a22628f97d8c0b5cf064cR284. In short, now allow large handshake messages (necessary to pass bogo tests), but after a large handshake message (larger than our 18k limit) we will shrink the buffer to a much smaller size again. We can probably tweak this more but, again, I think we should probably leave that to future PRs.

Reply:
The use case I have in mind is occasional large application-level messages on a long-running connection that mostly receives rare small messages.

As a simplified example, imagine a chat server that mostly deals with small messages from its clients but occasionally clients send pictures or videos. If there is no eventual auto-shrinking for  `self.buf`  then most of the connections will end up with `OpaqueMessage::MAX_WIRE_SIZE` buffers.

Nevertheless, I agree that finding/implementing the best strategy for auto-shrinking  is better left for a separate PR.

Reply:
My goal is still to aim for a zero-allocation API, so in that sense I don't want to spend a lot of time on optimizing the internal buffering strategy. But if you want to propose something after this is merged, we can definitely consider it.